File: android/app/src/main/java/com/vxin/app/data/repository/ChatRepository.kt
package com.vxin.app.data.repository

import com.vxin.app.core.realtime.ReactionEvent
import com.vxin.app.core.realtime.ReadEvent
import com.vxin.app.core.realtime.RedPacketClaimedEvent
import com.vxin.app.core.media.ChunkUploader
import com.vxin.app.core.media.MediaUploader
import com.vxin.app.core.realtime.SocketManager
import com.vxin.app.core.realtime.SocketStatus
import com.vxin.app.core.realtime.TypingEvent
import com.vxin.app.data.api.MessageApi
import com.vxin.app.data.model.Conversation
import com.vxin.app.data.model.DeleteMessageBody
import com.vxin.app.data.model.MarkReadRequest
import com.vxin.app.data.model.Message
import com.vxin.app.data.model.ReactBody
import okhttp3.MultipartBody
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ChatRepository @Inject constructor(
    private val api: MessageApi,
    private val chunkUploader: ChunkUploader,
    private val socketManager: SocketManager,
) {
    /** 实时连接状态（供 UI 显示「连接中/已连接」） */
    val socketStatus: StateFlow<SocketStatus> = socketManager.status

    /** 全局新消息流（各会话共用，UI 自行按 conversation_id 过滤） */
    val incomingMessages: SharedFlow<Message> = socketManager.incomingMessages

    /** typing / 已读 / 未读清除 事件流 */
    val typingEvents: SharedFlow<TypingEvent> = socketManager.typingEvents
    val readEvents: SharedFlow<ReadEvent> = socketManager.readEvents
    val unreadClearedEvents: SharedFlow<String> = socketManager.unreadClearedEvents
    val newConversationEvents: SharedFlow<Unit> = socketManager.newConversationEvents
    val messageDeletedEvents: SharedFlow<String> = socketManager.messageDeletedEvents
    val messageVanishedEvents: SharedFlow<String> = socketManager.messageVanishedEvents
    val batchDeletedEvents: SharedFlow<List<String>> = socketManager.messagesBatchDeleted
    val conversationClearedEvents: SharedFlow<String> = socketManager.conversationClearedEvents
    val reactionEvents: SharedFlow<ReactionEvent> = socketManager.reactionEvents
    val redPacketClaimedEvents: SharedFlow<RedPacketClaimedEvent> = socketManager.redPacketClaimedEvents
    val pinChangedEvents: SharedFlow<String> = socketManager.pinChangedEvents
    val groupGoneEvents: SharedFlow<String> = socketManager.groupGoneEvents
    val groupChangedEvents: SharedFlow<String> = socketManager.groupChangedEvents
    val messageEditedEvents: SharedFlow<com.vxin.app.core.realtime.MessageEditedEvent> = socketManager.messageEditedEvents
    /** 后台功能开关实时更新 → 最新 Features */
    val configUpdatedEvents: SharedFlow<com.vxin.app.data.model.Features> = socketManager.configUpdatedEvents

    fun joinConversation(conversationId: String) = socketManager.joinConversation(conversationId)
    fun emitTyping(conversationId: String) = socketManager.emitTyping(conversationId)
    fun emitStopTyping(conversationId: String) = socketManager.emitStopTyping(conversationId)

    /** 拍一拍（私聊可省略 targetId，服务端自动取对方） */
    fun nudge(conversationId: String, targetId: String? = null) = socketManager.emitNudge(conversationId, targetId)

    /** 设置/清除聊天专属背景（空串=清除） */
    suspend fun setConversationBackground(conversationId: String, background: String) =
        api.setBackground(conversationId, com.vxin.app.data.model.BackgroundBody(background))

    /** 标记会话已读 */
    suspend fun markRead(conversationId: String, messageId: String?) {
        runCatching { api.markRead(conversationId, MarkReadRequest(messageId)) }
    }

    suspend fun loadConversations(): List<Conversation> = api.conversations()

    suspend fun loadHistory(conversationId: String, before: Long? = null): List<Message> =
        api.history(conversationId, before = before)

    /** 会话内消息搜索（FTS5，倒序命中） */
    suspend fun searchInConversation(conversationId: String, q: String): List<Message> =
        api.searchInConversation(conversationId, q)

    suspend fun sendText(
        conversationId: String,
        content: String,
        replyToId: String? = null,
        clientMsgId: String? = null,
        scope: com.vxin.app.core.storage.MessageScope? = null,
    ): Result<Message> =
        socketManager.sendMessage(conversationId, content, replyToId, clientMsgId, scope)

    /** 上传媒体并返回服务端创建的消息（同时会经 Socket 广播给其他端） */
    suspend fun uploadMedia(conversationId: String, part: MultipartBody.Part): Message =
        api.upload(conversationId, part)

    /**
     * 按文件大小自动选择单次上传或分片上传（>8MB 走分片，对齐 Web）。
     */
    suspend fun uploadPrepared(
        conversationId: String,
        prepared: MediaUploader.Prepared,
        replyToId: String? = null,
        onProgress: ((Int) -> Unit)? = null,
    ): Message {
        return if (prepared.file.length() > ChunkUploader.CHUNK_THRESHOLD) {
            chunkUploader.upload(
                conversationId,
                prepared.file,
                prepared.displayName,
                prepared.mime,
                replyToId,
                onProgress,
            )
        } else {
            api.upload(conversationId, prepared.part)
        }
    }

    /** 撤回/删除消息 */
    suspend fun deleteMessage(msgId: String, forEveryone: Boolean = true) =
        runCatching { api.deleteMessage(msgId, DeleteMessageBody(forEveryone)) }

    /** 批量撤回/删除（多选） */
    suspend fun batchDelete(conversationId: String, msgIds: List<String>) =
        api.batchDelete(com.vxin.app.data.model.BatchDeleteBody(msgIds, conversationId))

    suspend fun vanishMessage(msgId: String) =
        runCatching { api.deleteMessage(msgId, DeleteMessageBody(vanish = true)) }

    /** 表情回应(切换) */
    suspend fun react(msgId: String, emoji: String) =
        runCatching { api.react(msgId, ReactBody(emoji)) }

    // ── 群置顶消息 ──
    suspend fun pinMessage(conversationId: String, msgId: String) =
        api.pinMessage(conversationId, com.vxin.app.data.model.PinMessageBody(msgId))

    suspend fun unpinMessage(conversationId: String, msgId: String) =
        api.unpinMessage(conversationId, msgId)

    suspend fun pinnedMessages(conversationId: String) = api.pinnedMessages(conversationId)

    // ── 会话操作 ──
    suspend fun setConversationPinned(conversationId: String, pinned: Boolean) =
        api.pinConversation(conversationId, com.vxin.app.data.model.PinConversationBody(if (pinned) 1 else 0))

    suspend fun setConversationMuted(conversationId: String, muted: Boolean) =
        api.muteConversation(conversationId, com.vxin.app.data.model.MuteConversationBody(if (muted) 1 else 0))

    /** 标为未读 */
    suspend fun markConversationUnread(conversationId: String) = api.markUnread(conversationId)

    /** 阅后即焚（seconds=0 关闭） */
    suspend fun setBurnAfter(conversationId: String, seconds: Int) =
        api.setBurnAfter(conversationId, com.vxin.app.data.model.BurnAfterBody(seconds))

    /** 文件传输助手会话（获取或创建），返回 conversationId */
    suspend fun fileHelper(): String = api.fileHelper().conversationId

    suspend fun clearMessages(conversationId: String) = api.clearMessages(conversationId)

    suspend fun editMessage(msgId: String, content: String) =
        api.editMessage(msgId, com.vxin.app.data.model.EditMessageBody(content))

    suspend fun forward(msgId: String, conversationIds: List<String>) =
        api.forward(com.vxin.app.data.model.ForwardBody(msgId, conversationIds))

    suspend fun collectMessage(msgId: String) = api.collectMessage(msgId)

    /** 导出指定会话的聊天记录，返回 text/plain 纯文本内容（由 Screen 负责写文件）。 */
    suspend fun exportConversation(conversationId: String): String =
        api.exportConversation(conversationId).string()

    // ── 功能A2: 消息定时发送 ────────────────────────────────────────────────

    /** 创建定时消息（send_at 为 UNIX 秒，需 ≥15分钟后且 ≤30天，后端会二次校验） */
    suspend fun scheduleMessage(
        conversationId: String,
        content: String,
        sendAt: Long,
    ): com.vxin.app.data.model.ScheduledMessage =
        api.scheduleMessage(
            com.vxin.app.data.model.ScheduleMessageBody(
                conversation_id = conversationId,
                content = content,
                send_at = sendAt,
            )
        )

    /** 我的全部定时消息列表（后端仅返回 pending 状态） */
    suspend fun scheduledMessages(): List<com.vxin.app.data.model.ScheduledMessage> =
        api.scheduledMessages()

    /** 取消指定定时消息（仅本人且 pending 状态可取消） */
    suspend fun cancelScheduledMessage(id: String) = api.cancelScheduledMessage(id)

    // ── 功能A2: @我消息聚合 ─────────────────────────────────────────────────

    /** 获取 @我 消息分页列表 */
    suspend fun mentionsMe(offset: Int = 0, limit: Int = 20): com.vxin.app.data.model.MentionsResponse =
        api.mentionsMe(offset = offset, limit = limit)

    // ── 功能A3: 聊天文件聚合视图 ─────────────────────────────────────────────

    /** 获取会话文件聚合分页列表（type=all|image|video|file） */
    suspend fun conversationFiles(
        conversationId: String,
        type: String = "all",
        offset: Int = 0,
        limit: Int = 30,
    ): com.vxin.app.data.model.ConversationFilesResponse =
        api.conversationFiles(conversationId, type = type, offset = offset, limit = limit)

    // ── 功能A3: 语音转文字 ──────────────────────────────────────────────────

    /** 语音转文字（幂等由后端管理；ASR 不可用后端返回 503） */
    suspend fun transcribe(msgId: String): com.vxin.app.data.model.TranscribeResponse =
        api.transcribe(msgId)
}

File: ios/Vxin/Data/Repositories/ChatRepository.swift
import Foundation
import Combine

/// 聊天仓库。与 Android ChatRepository 等价。
final class ChatRepository {
    static let shared = ChatRepository()
    private init() {}

    private let api = APIClient.shared
    private let socket = SocketService.shared

    /// 实时连接状态（供 UI 显示连接中/已连接）
    var statusPublisher: AnyPublisher<SocketStatus, Never> { socket.status.eraseToAnyPublisher() }

    /// 全局新消息流（各会话共用，UI 自行按 conversationId 过滤）
    var incomingPublisher: AnyPublisher<Message, Never> { socket.incoming.eraseToAnyPublisher() }

    var typingPublisher: AnyPublisher<TypingEvent, Never> { socket.typing.eraseToAnyPublisher() }
    var readPublisher: AnyPublisher<ReadEvent, Never> { socket.read.eraseToAnyPublisher() }
    var unreadClearedPublisher: AnyPublisher<String, Never> { socket.unreadCleared.eraseToAnyPublisher() }
    var newConversationPublisher: AnyPublisher<Void, Never> { socket.newConversation.eraseToAnyPublisher() }
    var messageDeletedPublisher: AnyPublisher<String, Never> { socket.messageDeleted.eraseToAnyPublisher() }
    var messageVanishedPublisher: AnyPublisher<String, Never> { socket.messageVanished.eraseToAnyPublisher() }
    var batchDeletedPublisher: AnyPublisher<[String], Never> { socket.batchDeleted.eraseToAnyPublisher() }
    var conversationClearedPublisher: AnyPublisher<String, Never> { socket.conversationCleared.eraseToAnyPublisher() }
    var reconnectedPublisher: AnyPublisher<Void, Never> { socket.reconnected.eraseToAnyPublisher() }
    var reactionPublisher: AnyPublisher<(String, [MessageReaction]), Never> { socket.reaction.eraseToAnyPublisher() }
    var redPacketClaimedPublisher: AnyPublisher<(String, String, Int), Never> { socket.redPacketClaimed.eraseToAnyPublisher() }
    var pinChangedPublisher: AnyPublisher<String, Never> { socket.pinChanged.eraseToAnyPublisher() }
    var groupGonePublisher: AnyPublisher<String, Never> { socket.groupGone.eraseToAnyPublisher() }
    var groupChangedPublisher: AnyPublisher<String, Never> { socket.groupChanged.eraseToAnyPublisher() }
    var messageEditedPublisher: AnyPublisher<(String, String, String), Never> { socket.messageEdited.eraseToAnyPublisher() }
    /// 被 @ 提及 → (conversationId, msgId)
    var mentionedPublisher: AnyPublisher<(convId: String, msgId: String), Never> { socket.mentioned.eraseToAnyPublisher() }

    func joinConversation(_ id: String) { socket.joinConversation(id) }
    func emitTyping(_ id: String) { socket.emitTyping(id) }
    func emitStopTyping(_ id: String) { socket.emitStopTyping(id) }

    /// 拍一拍（私聊可省略 targetId，服务端自动取对方）
    func nudge(conversationId: String, targetId: String? = nil) {
        socket.emitNudge(conversationId: conversationId, targetId: targetId)
    }

    /// 设置/清除聊天专属背景（空串=清除）
    func setConversationBackground(_ conversationId: String, background: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/background", method: "PUT", body: BackgroundBody(background: background)
        )
    }

    func loadConversations() async throws -> [Conversation] {
        try await api.send("api/messages/conversations")
    }

    func loadHistory(_ conversationId: String, before: Double? = nil) async throws -> [Message] {
        var path = "api/messages/\(conversationId)?limit=50"
        if let before { path += "&before=\(Int(before))" }
        return try await api.send(path)
    }

    /// 会话内消息搜索（FTS5，倒序命中）
    func searchInConversation(_ conversationId: String, q: String) async throws -> [Message] {
        let enc = q.addingPercentEncoding(withAllowedCharacters: .urlQueryValueAllowed) ?? q
        return try await api.send("api/messages/conversation/\(conversationId)/search?q=\(enc)")
    }

    func sendText(conversationId: String, content: String, replyToId: String? = nil) async -> Result<Message, Error> {
        await socket.sendMessage(conversationId: conversationId, content: content, replyToId: replyToId)
    }

    /// 撤回/删除消息
    func deleteMessage(_ msgId: String, forEveryone: Bool = true) async {
        let _: EmptyResponse? = try? await api.send(
            "api/messages/\(msgId)", method: "DELETE", body: DeleteMessageBody(forEveryone: forEveryone, vanish: nil)
        )
    }

    /// 彻底删除不留痕迹
    func vanishMessage(_ msgId: String) async {
        let _: EmptyResponse? = try? await api.send(
            "api/messages/\(msgId)", method: "DELETE", body: DeleteMessageBody(forEveryone: false, vanish: true)
        )
    }

    /// 表情回应(切换)
    func react(_ msgId: String, emoji: String) async -> [MessageReaction] {
        let resp: ReactResponse? = try? await api.send(
            "api/messages/\(msgId)/react", method: "POST", body: ReactBody(emoji: emoji)
        )
        return resp?.reactions ?? []
    }

    /// 上传媒体（图片/语音/文件）；返回服务端创建的消息（同时经 Socket 广播给其他端）
    func uploadMedia(conversationId: String, data: Data, fileName: String, mimeType: String) async throws -> Message {
        try await api.upload("api/messages/\(conversationId)/upload", fileData: data, fileName: fileName, mimeType: mimeType)
    }

    /// 标记会话已读（服务端发 message_read 给房间、sync:unread_cleared 给本人各端）
    func markRead(conversationId: String, messageId: String?) async {
        let _: EmptyResponse? = try? await api.send(
            "api/messages/conversation/\(conversationId)/read",
            method: "POST",
            body: MarkReadBody(messageId: messageId)
        )
    }

    // ── 群置顶消息 ──
    func pinMessage(conversationId: String, msgId: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/pin-message", method: "POST", body: PinMessageBody(msgId: msgId)
        )
    }

    func unpinMessage(conversationId: String, msgId: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/pin-message/\(msgId)", method: "DELETE"
        )
    }

    func pinnedMessages(conversationId: String) async throws -> [PinnedMessage] {
        try await api.send("api/messages/conversation/\(conversationId)/pinned-messages")
    }

    // ── 会话操作 ──
    func setConversationPinned(_ conversationId: String, pinned: Bool) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/pin", method: "POST", body: PinConvBody(pinned: pinned ? 1 : 0)
        )
    }

    func setConversationMuted(_ conversationId: String, muted: Bool) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/mute", method: "POST", body: MuteConvBody(muted: muted ? 1 : 0)
        )
    }

    func clearMessages(_ conversationId: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/conversation/\(conversationId)/messages", method: "DELETE"
        )
    }

    func editMessage(_ msgId: String, content: String) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/\(msgId)/edit", method: "PUT", body: EditBody(content: content)
        )
    }

    func forward(msgId: String, conversationIds: [String]) async throws {
        let _: EmptyResponse = try await api.send(
            "api/messages/forward", method: "POST", body: ForwardBody(msgId: msgId, conversationIds: conversationIds)
        )
    }

    func collectMessage(_ msgId: String) async throws {
        let _: EmptyResponse = try await api.send("api/messages/\(msgId)/collect", method: "POST")
    }
}

private struct MarkReadBody: Encodable { let messageId: String? }
private struct BackgroundBody: Encodable { let background: String }
private struct PinMessageBody: Encodable { let msgId: String }
private struct PinConvBody: Encodable { let pinned: Int }
private struct MuteConvBody: Encodable { let muted: Int }
private struct EditBody: Encodable { let content: String }
private struct ForwardBody: Encodable { let msgId: String; let conversationIds: [String] }
private struct DeleteMessageBody: Encodable { let forEveryone: Bool; let vanish: Bool? }
private struct ReactBody: Encodable { let emoji: String }
private struct ReactResponse: Decodable { let reactions: [MessageReaction] }
